### ISSUE TITLE
[CT-3253] [Bug] Failure to load and process a partial parse file results in a log with a malformed JSON dump

### ISSUE THREAD

**Issue by will-sargent-dbtlabs** [dbt-labs/dbt-core#8888]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

In testing partial parsing behavior, I was testing the scenario of moving from version 1.5 to version 1.7.

This resulted in an error when core attempted to use a `partial_parse.msgpack` from 1.5, that contained legacy metrics layer things that do not translate to 1.7.

The error message, including the "offending" manifest JSON blob, was dumped into the debug logs, which is great.

The problem is that the JSON that was dumped into the logs, is malformed in multiple ways, in that you cannot copy-paste it, and format it in VS Code or any other JSON viewer/formatter to make it human readable.

Note that my particular issue is only related to dbt Cloud running core.

### Expected Behavior

The log line JSON dump, should be pasteable into a JSON viewer/formatter, and be able to be formatted for analysis.

### Steps To Reproduce

1. Requires a dbt Cloud account with partial parsing method update activated.
2. Internally, a repo called `ia-performance-testing` is where I sourced the problem from.
3. A model called `export_the_number_one` is an excellent, single select model to do a build against, as it is self-contained:
`dbt build --select export_the_number_one`
4. Do a run in 1.5 against a project with SL code in it that is incompatible with 1.5. It doesn't matter if the run fails or succeeds, as long as it completes the parsing phase.
5. Change the env to 1.7
6. Re run the job
7. See the error in the debug logs
8. Attempt to copy-paste the JSON into a formatting tool
9. Errors from formatter

### Relevant log output

```shell
Internal dbt Labs only link to log file example:
https://drive.google.com/file/d/12rZ9BMlNvKR1ej7xkLjdDMSVLhInvfSn/view?usp=sharing
```

### Environment

```markdown
dbt Cloud runs 1.5 and 1.7
```

### Which database adapter are you using with dbt?

snowflake

**Comment by emmyoop** [comment]
There some useful context to reproduce this in [this internal slack thread](https://dbt-labs.slack.com/archives/C05FWBP9X1U/p1698159803460469).

**Comment by dbeatty10** [comment]
Thanks for opening @will-sargent-dbtlabs !

### Parsing dictionary output

That error output looks like JSON... but it's actually the string representation of a Python dictionary! So that's why you were having trouble formatting it. Expand the toggle below to see a Python script that can read in a file with that format and convert it to an actual Python `dict`.

<details>
<summary>Click to toggle `parse_dict.py`</summary>

#### Convert a python `str` into a `dict`

1\. Put your string representation of a python dictionary into a file named `py_dict_str.txt`.

2\. Create this python script with the name `parse_dict.py`

```python
import ast


def read_file(file_path):
    """
    Read the contents of a file.

    :param file_path: Path to the file
    :return: String contents of the file or None if an error occurred
    """
    try:
        with open(file_path, 'r') as file:
            content = file.read()
            return content
    except IOError as e:
        print(f"Error reading from file: {e}")
        return None


def string_to_dict(dict_string):
    """
    Convert a string representation of a dictionary into a dictionary.

    This function uses ast.literal_eval() which is considered
    safe as it does not execute functions or operations within
    the string, avoiding code injection issues that the eval()
    function is known for.

    :param dict_string: String representation of a dictionary
    :return: Dictionary object
    """
    try:
        # Convert the string to a dictionary
        result_dict = ast.literal_eval(dict_string)
        return result_dict
    except ValueError as e:
        print(f"Error: {e}")
        return {}
    except SyntaxError as e:
        print(f"Error: {e}")
        return {}


# Example usage
if __name__ == "__main__":

    # Specify the path to your file
    file_path = "py_dict_str.txt"

    # Read the string from the file
    dict_string = read_file(file_path)

    # Convert the string to an actual dictionary
    my_dict = string_to_dict(dict_string)
    
    # Print the dictionary contents
    print(my_dict)
```

3\. Run the script

```shell
python parse_dict.py
```

</details>

### Digging in

@emily identified [this](https://github.com/dbt-labs/dbt-core/blob/35f46dac8ce512dfa9174316c8d9ac0c0bc477b6/core/dbt/parser/manifest.py#L896) as the portion of the code that is emitting this particular error message.

Re-labeling this to a feature request since dbt-core isn't doing anything erroneous.

But as a feature request, I'm not seeing anything obvious we should do here. Truncating the result would actually be _worse_ in my opinion, because then you can't see the full context that caused the error (not that full context is actually helpful in this particular case 😅).

This is a [`ParsedFileLoadFailed`](https://github.com/dbt-labs/dbt-core/blob/35f46dac8ce512dfa9174316c8d9ac0c0bc477b6/core/dbt/events/types.py#L905) error, which is a debug-level event.

Looking at your logs, `debug=True`:
```
2023-10-23 15:58:37.949616 (MainThread): 15:58:37  running dbt with arguments {'printer_width': '80', 'indirect_selection': 'eager', 'write_json': 'True', 'log_cache_events': 'False', 'partial_parse': 'True', 'cache_selected_only': 'False', 'profiles_dir': '/tmp/jobs/122/.dbt', 'version_check': 'False', 'warn_error': 'None', 'log_path': '/tmp/jobs/122/target/logs', 'fail_fast': 'False', 'debug': 'True', 'use_colors': 'True', 'use_experimental_parser': 'False', 'no_print': 'None', 'quiet': 'False', 'warn_error_options': 'WarnErrorOptions(include=[], exclude=[])', 'static_parser': 'True', 'log_format': 'json', 'introspect': 'True', 'target_path': 'None', 'invocation_command': 'dbt --log-format json --debug build --select export_the_number_one --no-version-check --target default --profile user --profiles-dir /tmp/jobs/122/.dbt', 'send_anonymous_usage_stats': 'True'}
```

If the [`DEBUG` config](https://docs.getdbt.com/reference/global-configs/logs#debug-level-logging) is set to `False`, then it won't be so verbose like this.

### Summary

Since we've gotten to the bottom of how to format the log output _and_ how to remove this verbose output from the logs emitted to stdout, I'm going to close this as "not planned".

We can re-open this if there's some key consideration I accidentally overlooked.

**Comment by will-sargent-dbtlabs** [comment]
Thanks @dbeatty10 , that was a very helpful unlock!
I also found this plugin for VSC which made swift work of it:
https://marketplace.visualstudio.com/items?itemName=mgesbert.indent-nested-dictionary

I guess my ask would be: Since if this trips me up, most folks at home have no chance, should we be converting this to JSON instead so that users with the typical dbt Cloud skillset can untangle it as JSON, and fix the error?

In general, I guess that could be said for any core error that returns a python dict instead of JSON.  I'm opinionated towards making this easy for end users to pick apart and figure out what is broken, but there might be other legitimate reasons that converting this to JSON as a rule is actually less helpful?

Sample output:
<img width="961" alt="image" src="https://github.com/dbt-labs/dbt-core/assets/109557847/eabb39c1-9d97-4194-8c7e-dccf9cd7c5d8">

**Comment by dbeatty10** [comment]
The result of running that VSC extension is beautiful 😍

💯 agreed that we want to make it as hard as possible to get broken in the first place and as easy as possible to recover.

### TLDR
If dbt Cloud users are running into dbt errors that have insufficient error messages coming from dbt-core, then we can absolutely assess those on a case-by-case basis. Please raise issues as needed!

But I'm not seeing anything in this _particular_ case in which we'd be keen to do anything differently.

### This particular case

In this case, did dbt actually fail to execute? Looking at the logs [here](https://drive.google.com/file/d/12rZ9BMlNvKR1ej7xkLjdDMSVLhInvfSn/view), it looks like it completed successfully (even though it wasn't able to partial parse):
```
2023-10-23 15:59:50.833748 (MainThread): 15:59:50  Finished running 1 view model, 1 hook in 0 hours 0 minutes and 6.95 seconds (6.95s).
2023-10-23 15:59:50.834902 (MainThread): 15:59:50
2023-10-23 15:59:50.834902 (MainThread): 15:59:50  Command end result
2023-10-23 15:59:52.034527 (MainThread): 15:59:52  
2023-10-23 15:59:52.035322 (MainThread): 15:59:52
2023-10-23 15:59:52.035322 (MainThread): 15:59:52   [32mCompleted successfully [0m
2023-10-23 15:59:52.035889 (MainThread): 15:59:52  
2023-10-23 15:59:52.036557 (MainThread): 15:59:52
2023-10-23 15:59:52.036557 (MainThread): 15:59:52  Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

Partial parsing is a nice-to-have performance optimization, so it might be "okay" that it didn't work in this scenario. If not, then we should open up a bug report for it.

In general, we'd also consider it acceptable if debug logs have a higher barrier to entry since they necessarily have a lot more (and sometimes arcane) detail.

### Reasons against trying to convert content of error messages to JSON in general

While basic serialization from Python objects to JSON is often simple, non-trivial use cases require a good understanding of both the data being serialized and the quirks of the JSON format itself.

Here are just a few potential issues that can arise:
- non-serializable types
- circular references
- handling of tuples and sets
- large or complex nested objects
- non-string keys in dictionaries
- binary data

If any of those things goes wrong during serialization, then we can end up accidentally triggering an exception while trying to handle the original exception, further obscuring what's going on.

So it's not just as simple as converting from Python to JSON.

And even if we could do those conversions simply and reliably, we're frequently just surfacing errors raised from other Python libraries that we don't manage. So whatever information they provide is what we log out.

**Comment by will-sargent-dbtlabs** [comment]
Thanks @dbeatty10 for the further write-up, and I agree: at the end of the day, the run didn't fail, so there isn't a real action here. 

Also thank you for highlighting the difficulties of the Dict to JSON conversion. Makes sense, and I suspected it might actually make a bigger mess.  The better option is, if we need to investigate in the future, we have an existing tool that can format things into readable, and go from there.  No extra required.

Always enjoying the journey of learning here!  Thanks again!
